### ISSUE TITLE
DeflateCRC32Stream is broken with NodeJS 15.6.0

### ISSUE THREAD

**Issue by xfournet** [archiverjs/node-crc32-stream#31]
Starting with NodeJS 15.6.0, calling `end(chunk)`  on a `DeflateCRC32Stream` instance, the checksum is no more updated. This cause this bug https://github.com/archiverjs/node-archiver/issues/491

```
const { DeflateCRC32Stream } = require('crc32-stream');

console.log('Node version:', process.version);
console.log();

const s = new DeflateCRC32Stream();
s.end(Buffer.from('hello world'));
console.log('crc32: ', s.hex());
console.log('rawSize: ', s.size(false));
```

```
Node version: v15.5.1

crc32:  0D4A1185
rawSize:  11
```

```
Node version: v15.6.0

crc32:  00000000
rawSize:  0
```

**Comment by Aaron1011** [comment]
Looking at the changelog for Ndoe v15.6.0, I noticed https://github.com/nodejs/node/pull/36618, which seems potentially related.

**Comment by xfournet** [comment]
Probably the end function was calling the write (overloaded in DeflateCRC32Stream) but now end is not calling this write function anymore causing the problem.

**Comment by Aaron1011** [comment]
Minimized even further:

```node
const {DeflateRaw} = require('zlib');

class MyStream extends DeflateRaw {
    constructor(options) {
        super(options);
    }

    write(chunk, enc, cb) {
        console.log("Write: " + chunk + " " + enc + " " + cb);
    }
}

console.log("Running");
const s = new MyStream();
s.end(Buffer.from('hello world'));
```

On Node v15.5.0, this prints:

```
Running
Write: hello world undefined undefined
```

On Node v15.6.0, this prints:

```
Running
```

**Comment by Aaron1011** [comment]
Opened https://github.com/nodejs/node/issues/37027

**Comment by xfournet** [comment]
In meantime I wrote this patch (a bit ugly) to use with [patch-package](https://www.npmjs.com/package/patch-package). It works with NodeJS 15.5.1 & 15.6.0

File `patches/crc32-stream+4.0.1.patch`
```
diff --git a/node_modules/crc32-stream/lib/deflate-crc32-stream.js b/node_modules/crc32-stream/lib/deflate-crc32-stream.js
index ffe7661..65bf095 100644
--- a/node_modules/crc32-stream/lib/deflate-crc32-stream.js
+++ b/node_modules/crc32-stream/lib/deflate-crc32-stream.js
@@ -12,6 +12,15 @@ const {DeflateRaw} = require('zlib');
 
 const crc32 = require('crc-32');
 
+class TestDeflateRaw extends DeflateRaw {
+  write(chunk, enc, cb) {
+    this.writeCalled = true;
+  }
+}
+const testStream = new TestDeflateRaw();
+testStream.end(Buffer.from('testStream'));
+const processEndChunk = !testStream.writeCalled;
+
 class DeflateCRC32Stream extends DeflateRaw {
   constructor(options) {
     super(options);
@@ -40,6 +49,15 @@ class DeflateCRC32Stream extends DeflateRaw {
     return super.write(chunk, enc, cb);
   }
 
+  end(chunk, enc, cb) {
+    if (processEndChunk && chunk) {
+      this.checksum = crc32.buf(chunk, this.checksum) >>> 0;
+      this.rawSize += chunk.length;
+    }
+
+    return super.end(chunk, enc, cb);
+  }
+
   digest(encoding) {
     const checksum = Buffer.allocUnsafe(4);
     checksum.writeUInt32BE(this.checksum >>> 0, 0);
```

**Comment by xfournet** [comment]
Fix in PR #32
